DOCS: Update Gamepad documentation

jfreire-unity opened the pull request:
### Description

## Documentation fixes
Updates Gamepad documentation based on Unity Package API Quality Standards by addressing Score Explanation issues:
- Fix related links for all, buttonEast, buttonNorth, buttonSouth, buttonWest, current
- Removed value tag for aButton, all, bButton, buttonEast, buttonNorth, buttonSouth, buttonWest, circleButton, crossButton, current, dpad, leftShoulder, leftStick, leftStickButton, leftTrigger, rightShoulder, rightStick, rightStickButton, rightTrigger, selectButton, squareButton, startButton, triangleButton, xButton, yButton.
- Fixed insufficient remarks length (threshold is 120) on MakeCurrent()
- Added related links (required for this API type) to Gamepad class
- Added  code samples (required for this API type) each for MakeCurrent(), PauseHaptics(), ResetHaptics(), ResumeHaptics(), SetMotorSpeeds(float, float)
  - ❌ Decided not to add examples to FinishSetup(),  OnAdded(), OnRemoved() since they are protected methods.
- Added remarks  for OnAdded(), OnRemoved(), PauseHaptics(), ResetHaptics(), ResumeHaptics()
- Expanded code samples (required for this API type) for Gamepad class.

## Reusability of code examples improvement

This PR follows the approach mentioned in https://docs.unity.com/internal/content-ops/en-us/documentation/content-model/code-snippets-examples.

Many packages such as the Localization and Entities packages follow this convention. I'm throwing the first rock for our package :D 

Also, see the [Slack thread](https://unity.slack.com/archives/C08100WBU92/p1733326529294579) for more insights in this approach, so that I could fix PVP problems.

I'm open for deeper improvements in follow-up PRs but this seems like a good way forward for now.
Let me know your thoughts. 

### Testing status & QA

Docs build on my machine.

### Overall Product Risks

None

- Complexity: Minimal
- Halo Effect: Minimal

### Comments to reviewers

Please let me know of any inconsistencies and improvements I should do.
Also, let me know if the approach to reuse code samples in the folder `DocCodeSamples` is ok.
### Checklist

Before review:

- [x] Docs for new/changed API's.
    - Xmldoc cross references are set correctly.
    - Added explanation how the API works.
    - Usage code examples added.
    - The manual is updated, if needed.

During merge:

- [x] Commit message for squash-merge is prefixed with one of the list:
    - `NEW: ___`.
    - `FIX: ___`.
    - `DOCS: ___`.
    - `CHANGE: ___`.
    - `RELEASE: 1.1.0-preview.3`.

After merge:

- [ ] Create forward/backward port if needed. If you are blocked from creating a forward port now please add a task to ISX-1444.

jfreire-unity committed:
Do documentation improvements to Gamepad.cs

Addresses most issues to be solved in Docs Quality Week 2024 December

jfreire-unity committed:
Make DualMotorRumble.cs more consistent based on Gamepad.cs changes.

stefanunity commented on Packages/com.unity.inputsystem/InputSystem/Devices/Gamepad.cs:
```diff
@@ -132,20 +131,18 @@ public struct GamepadState : IInputStateTypeInfo
+        /// Right stick position.
         /// </summary>
-        /// <value>Right stick position.</value>
+        /// <remarks>Each axis from -1 to 1 with 0 being center position.</remarks>
```
remarks for left and right stick axis are not the same

stefanunity commented on Packages/com.unity.inputsystem/InputSystem/Devices/Gamepad.cs:
```diff
@@ -402,106 +399,146 @@ namespace UnityEngine.InputSystem
+    ///
+    ///         // Read left stick value and perform some code based on the value
+    ///         Vector2 move = gamepad.leftStick.ReadValue();
+    ///         {
```
why { } if there is no control statement?

jfreire-unity replied:
It's just to scope out  the`move` variable, as seen in an example [here](https://docs.unity3d.com/Packages/com.unity.inputsystem@1.11/manual/Workflow-Direct.html)

jfreire-unity committed:
Right stick remarks fix based on review

duckets committed:
Lengthened some doc fields to meet minimum length requirements

ekcoh commented on Packages/com.unity.inputsystem/InputSystem/Devices/Gamepad.cs:
```diff
@@ -174,7 +171,7 @@ public struct GamepadState : IInputStateTypeInfo
         /// State format tag for GamepadState.
         /// </summary>
-        /// <value>Returns "GPAD".</value>
+        /// <remarks>Returns "GPAD".</remarks>
```
I would guess this violates length requirement, is a remark required here? Could it be "State format tag for GamepadState ("GPAD")."

jfreire-unity replied:
Fixed

ekcoh commented on Packages/com.unity.inputsystem/InputSystem/Devices/Gamepad.cs:
```diff
@@ -402,106 +399,146 @@ namespace UnityEngine.InputSystem
+    ///             gamepad.ResetHaptics();
+    ///         }
+    ///     }
+    /// }
```
I am afraid example nested inside remarks will generate CI xmldcc error?

jfreire-unity replied:
Ah didn't notice it, will fix

ekcoh commented on Packages/com.unity.inputsystem/InputSystem/Devices/Gamepad.cs:
```diff
@@ -402,106 +399,146 @@ namespace UnityEngine.InputSystem
         /// <remarks>
-        /// On an Xbox controller, this is the Y button and on the PS4 controller, this is the
-        /// triangle button.
+        /// Control representing the Y/Triangle face button.
```
Side-note: I think this doc clearly illustrates the problem with current aliased enums. E.g. imagine if we supported 800 gamepads, how many variants of the mapping should be mentioned here? :)

jfreire-unity replied:
Yeah, but I think this only used for convenience. The aliases could (or maybe should)  be removed in a 2.x

ekcoh commented on Packages/com.unity.inputsystem/InputSystem/Devices/Gamepad.cs:
```diff
@@ -730,9 +788,19 @@ protected override void OnRemoved()
         /// <seealso cref="IDualMotorRumble"/>
+        /// <example>
+        /// <code>
+        /// <inheritdoc cref="SetMotorSpeeds"/>
```
Nice one

ekcoh reviewed (changes requested):
Seems like there are more work to do unfortunately, see dry-run CI job:


PVP-150-1: UnityEngine.InputSystem.Gamepad: void OnAdded(): <inheritdoc> in block context (only allowed in top-level context)
PVP-150-1: UnityEngine.InputSystem.Gamepad: void OnRemoved(): <inheritdoc> in block context (only allowed in top-level context)
PVP-150-1: UnityEngine.InputSystem.Gamepad: void PauseHaptics(): <inheritdoc> in block context (only allowed in top-level context)
PVP-150-1: UnityEngine.InputSystem.Gamepad: void PauseHaptics(): tag inside <code>
PVP-150-1: UnityEngine.InputSystem.Gamepad: void ResumeHaptics(): <inheritdoc> in block context (only allowed in top-level context)
PVP-150-1: UnityEngine.InputSystem.Gamepad: void ResumeHaptics(): tag inside <code>
PVP-150-1: UnityEngine.InputSystem.Gamepad: void ResetHaptics(): <inheritdoc> in block context (only allowed in top-level context)
PVP-150-1: UnityEngine.InputSystem.Gamepad: void ResetHaptics(): tag inside <code>

jfreire-unity committed:
Add some more small fixes

Will follow up on errors next.

jfreire-unity commented:
> Seems like there are more work to do unfortunately, see dry-run CI job:
> 
> PVP-150-1: UnityEngine.InputSystem.Gamepad: void OnAdded(): in block context (only allowed in top-level context) PVP-150-1: UnityEngine.InputSystem.Gamepad: void OnRemoved(): in block context (only allowed in top-level context) PVP-150-1: UnityEngine.InputSystem.Gamepad: void PauseHaptics(): in block context (only allowed in top-level context) PVP-150-1: UnityEngine.InputSystem.Gamepad: void PauseHaptics(): tag inside `` PVP-150-1: UnityEngine.InputSystem.Gamepad: void ResumeHaptics(): in block context (only allowed in top-level context) PVP-150-1: UnityEngine.InputSystem.Gamepad: void ResumeHaptics(): tag inside ` PVP-150-1: UnityEngine.InputSystem.Gamepad: void ResetHaptics(): in block context (only allowed in top-level context) PVP-150-1: UnityEngine.InputSystem.Gamepad: void ResetHaptics(): tag inside ` ``

Missed those, will fix them as soon as I understand them :D

jfreire-unity committed:
WIP Attempt to fix CI errors

ritamerkl commented on Packages/com.unity.inputsystem/InputSystem/Devices/Gamepad.cs:
```diff
@@ -154,18 +151,18 @@ public struct GamepadState : IInputStateTypeInfo
+        /// The current position of the left trigger on a gamepad.
         /// </summary>
-        /// <value>Position of left trigger.</value>
+        /// <remarks>The value's range goes from 0 to 1, where 0 represents the trigger not pressed at all, and 1 represents the trigger in its fully pressed position.</remarks>
```
not pressed at all -> in it's neutral position, its -> it's

ritamerkl replied:
likewise in other parts of the code.

jfreire-unity replied:
Fixed

ritamerkl commented on Packages/com.unity.inputsystem/InputSystem/Devices/Gamepad.cs:
```diff
@@ -637,28 +663,29 @@ public ButtonControl this[GamepadButton button]
         /// you need it. Whenever the gamepad setup changes, the value returned by this getter
         /// is invalidated.
+        ///
+        /// Alternately, if you want a single gamepad, you can use <see cref="current"/> for example.
```
"if you want a single gamepad" sounds a bit odd. maybe "for querying a single gamepad"?

jfreire-unity replied:
Thanks! Fixed

ritamerkl commented on Packages/com.unity.inputsystem/InputSystem/Devices/Haptics/DualMotorRumble.cs:
```diff
@@ -39,9 +39,13 @@ internal struct DualMotorRumble
         /// <summary>
-        /// Reset motor speeds to zero but retain current values for <see cref="lowFrequencyMotorSpeed"/>
-        /// and <see cref="highFrequencyMotorSpeed"/>.
+        /// Reset motor speeds to zero.
```
Is this summary long enough to fit the requirements in count of characters?

ritamerkl reviewed (approved):
LGTM, nice code examples

jfreire-unity committed:
Fix PVP errors and improve examples reusability

 <inheritdoc cref> can only be used as a top-level xml tag.
 Since it is not possible to inherit code examples, I went ahead with the guidelines for API Package Documentation.

jfreire-unity committed:
Improvements based on PR review

jfreire-unity commented:
> Seems like there are more work to do unfortunately, see dry-run CI job:
> 
> PVP-150-1: UnityEngine.InputSystem.Gamepad: void OnAdded(): in block context (only allowed in top-level context) PVP-150-1: UnityEngine.InputSystem.Gamepad: void OnRemoved(): in block context (only allowed in top-level context) PVP-150-1: UnityEngine.InputSystem.Gamepad: void PauseHaptics(): in block context (only allowed in top-level context) PVP-150-1: UnityEngine.InputSystem.Gamepad: void PauseHaptics(): tag inside `` PVP-150-1: UnityEngine.InputSystem.Gamepad: void ResumeHaptics(): in block context (only allowed in top-level context) PVP-150-1: UnityEngine.InputSystem.Gamepad: void ResumeHaptics(): tag inside ` PVP-150-1: UnityEngine.InputSystem.Gamepad: void ResetHaptics(): in block context (only allowed in top-level context) PVP-150-1: UnityEngine.InputSystem.Gamepad: void ResetHaptics(): tag inside ` ``

The `<inheritdoc />` tag needs to be used at the top-level (it can't be a in-block tag). I fixed that as best as possible.

ritamerkl commented on Packages/com.unity.inputsystem/InputSystem/Devices/Gamepad.cs:
```diff
@@ -399,109 +402,96 @@ namespace UnityEngine.InputSystem
-    /// // Rumble the left motor on the current gamepad slightly.
-    /// Gamepad.current.SetMotorSpeeds(0.2f, 0.
-    /// </code>
+    /// <code source="../../DocCodeSamples.Tests/GamepadExample.cs" />
```
cool! That is handy!

ritamerkl commented on Packages/com.unity.inputsystem/DocCodeSamples.Tests/GamepadExample.cs:
```diff
@@ -0,0 +1,50 @@
+        // Read left stick value and perform some code based on the value
+        Vector2 move = gamepad.leftStick.ReadValue();
+        {
+            // Do 'move' code here
```
Sounds a bit odd
// Use the Vector2 move for the game logic

ritamerkl reviewed (commented):
LGTM, only one minor comment - Despite CI seems to fail, approving to unblock you

jfreire-unity committed:
Add namespace and missing files, and improve comment

jfreire-unity committed:
Merge branch 'develop' into docs-quality-week-2024-gamepad

ekcoh reviewed (approved):
LGTM, nice to see the referenced doc class under asmdef

jfreire-unity committed:
Merge branch 'develop' into docs-quality-week-2024-gamepad

duckets committed:
added more detail for OnAdded and OnRemoved docs
